### ISSUE TITLE
KeyError: 'module_name' at evaluating shapenet

### ISSUE THREAD

**Issue by molnarszilard** [nicolas-chaulet/torch-points3d#480]
Hi. I have trained a model with shapenet as in the example (segmentation, pointnet2, pointnet2_charlesssg).
But when I am trying to run eval.py, after it finishes processing I get the following error:

Traceback (most recent call last):
  File "eval.py", line 16, in <module>
    main()
  File "/home/cuda/.local/lib/python3.6/site-packages/hydra/main.py", line 24, in decorated_main
    strict=strict,
  File "/home/cuda/.local/lib/python3.6/site-packages/hydra/_internal/utils.py", line 174, in run_hydra
    overrides=args.overrides,
  File "/home/cuda/.local/lib/python3.6/site-packages/hydra/_internal/hydra.py", line 86, in run
    job_subdir_key=None,
  File "/home/cuda/.local/lib/python3.6/site-packages/hydra/plugins/common/utils.py", line 109, in run_job
    ret.return_value = task_function(task_cfg)
  File "eval.py", line 11, in main
    trainer = Trainer(cfg)
  File "/home/cuda/torch-points3d/torch_points3d/trainer.py", line 47, in __init__
    self._initialize_trainer()
  File "/home/cuda/torch-points3d/torch_points3d/trainer.py", line 89, in _initialize_trainer
    self._dataset, weight_name=self._cfg.training.weight_name
  File "/home/cuda/torch-points3d/torch_points3d/metrics/model_checkpoint.py", line 183, in create_model
    model = instantiate_model(run_config, dataset)
  File "/home/cuda/torch-points3d/torch_points3d/models/model_factory.py", line 42, in instantiate_model
    model = model_cls(model_config, "dummy", dataset, modellib)
  File "/home/cuda/torch-points3d/torch_points3d/models/segmentation/pointnet2.py", line 36, in __init__
    UnetBasedModel.__init__(self, option, model_type, dataset, modules)
  File "/home/cuda/torch-points3d/torch_points3d/models/base_architectures/unet.py", line 89, in __init__
    self._init_from_compact_format(opt, model_type, dataset, modules_lib)
  File "/home/cuda/torch-points3d/torch_points3d/models/base_architectures/unet.py", line 114, in _init_from_compact_format
    args_up=args_up, args_innermost=opt.innermost, modules_lib=modules_lib, submodule=None, innermost=True,
  File "/home/cuda/torch-points3d/torch_points3d/models/base_architectures/unet.py", line 271, in __init__
    module_name = self.get_from_kwargs(args_innermost, "module_name")
  File "/home/cuda/torch-points3d/torch_points3d/models/base_architectures/unet.py", line 242, in get_from_kwargs
    kwargs.pop(name)
  File "/home/cuda/.local/lib/python3.6/site-packages/omegaconf/dictconfig.py", line 119, in pop
    raise KeyError(key)
KeyError: 'module_name'

**Comment by nicolas-chaulet** [comment]
Could you print the full log? Are you sure that the path to your model and your model name are correct in conf/eval.yaml?

**Comment by molnarszilard** [comment]
As far as I can tell the path and model name are correct in the eval file. I have even experimented with setting make_submission to True as you mentioned in https://github.com/nicolas-chaulet/torch-points3d/issues/433, but the result doesn't change.
This is the full log.
$ python eval.py
batch_size: 1
checkpoint_dir: /home/cuda/torch-points3d/outputs/2020-10-26/11-44-16
cuda: 0
enable_cudnn: true
enable_dropout: false
model_name: pointnet2_charlesssg
num_workers: 1
precompute_multi_scale: true
tracker_options:
  full_res: false
  make_submission: true
voting_runs: 1
weight_name: latest

[2020-10-29 13:48:14,117][torch_points3d.trainer][INFO] - DEVICE : cuda
[2020-10-29 13:48:14,117][torch_points3d.metrics.model_checkpoint][INFO] - Loading checkpoint from /home/cuda/torch-points3d/outputs/2020-10-26/11-44-16/pointnet2_charlesssg.pt
Processing...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 12137/12137 [00:00<00:00, 15365.23it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1870/1870 [00:00<00:00, 28834.46it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2874/2874 [00:00<00:00, 31671.01it/s]
Done!
Traceback (most recent call last):
  File "eval.py", line 16, in <module>
    main()
  File "/home/cuda/.local/lib/python3.6/site-packages/hydra/main.py", line 24, in decorated_main
    strict=strict,
  File "/home/cuda/.local/lib/python3.6/site-packages/hydra/_internal/utils.py", line 174, in run_hydra
    overrides=args.overrides,
  File "/home/cuda/.local/lib/python3.6/site-packages/hydra/_internal/hydra.py", line 86, in run
    job_subdir_key=None,
  File "/home/cuda/.local/lib/python3.6/site-packages/hydra/plugins/common/utils.py", line 109, in run_job
    ret.return_value = task_function(task_cfg)
  File "eval.py", line 11, in main
    trainer = Trainer(cfg)
  File "/home/cuda/torch-points3d/torch_points3d/trainer.py", line 47, in __init__
    self._initialize_trainer()
  File "/home/cuda/torch-points3d/torch_points3d/trainer.py", line 89, in _initialize_trainer
    self._dataset, weight_name=self._cfg.training.weight_name
  File "/home/cuda/torch-points3d/torch_points3d/metrics/model_checkpoint.py", line 183, in create_model
    model = instantiate_model(run_config, dataset)
  File "/home/cuda/torch-points3d/torch_points3d/models/model_factory.py", line 42, in instantiate_model
    model = model_cls(model_config, "dummy", dataset, modellib)
  File "/home/cuda/torch-points3d/torch_points3d/models/segmentation/pointnet2.py", line 36, in __init__
    UnetBasedModel.__init__(self, option, model_type, dataset, modules)
  File "/home/cuda/torch-points3d/torch_points3d/models/base_architectures/unet.py", line 89, in __init__
    self._init_from_compact_format(opt, model_type, dataset, modules_lib)
  File "/home/cuda/torch-points3d/torch_points3d/models/base_architectures/unet.py", line 114, in _init_from_compact_format
    args_up=args_up, args_innermost=opt.innermost, modules_lib=modules_lib, submodule=None, innermost=True,
  File "/home/cuda/torch-points3d/torch_points3d/models/base_architectures/unet.py", line 271, in __init__
    module_name = self.get_from_kwargs(args_innermost, "module_name")
  File "/home/cuda/torch-points3d/torch_points3d/models/base_architectures/unet.py", line 242, in get_from_kwargs
    kwargs.pop(name)
  File "/home/cuda/.local/lib/python3.6/site-packages/omegaconf/dictconfig.py", line 119, in pop
    raise KeyError(key)
KeyError: 'module_name'

**Comment by christianlandgraf** [comment]
I experienced the same problem with the only difference that I use Python3.8 and it already fails at the beginning before processing anything.
I noticed, that the loaded model config from the saved outputs is different to the actual yaml configuration. Here is the printed version:

`{'class': 'pointnet2.PointNet2_D', 'conv_type': 'DENSE', 'use_category': '${data.use_category}', 'down_conv': {'module_name': 'PointNetMSGDown', 'npoint': [512, 128], 'radii': [[0.2], [0.4]], 'nsamples': [[64], [64]], 'down_conv_nn': [[[6, 64, 64, 128]], [[131, 128, 128, 256]]]}, 'innermost': {'nn': [259, 256, 512, 1024]}, 'up_conv': {'module_name': 'DenseFPModule', 'up_conv_nn': [[1280, 256, 256], [384, 256, 128], [131, 128, 128, 128]], 'skip': True}, 'mlp_cls': {'nn': [129, 128], 'dropout': 0.5}}`

It seems as if the innermost module misses its module_name. The same is happening for pointnet2 insted of pointnet2_charlessg.
I tried to manually add the correct module in line 271 in unet.py
```
            module_name = self.get_from_kwargs(args_innermost, "module_name")
            # module_name = "GlobalDenseBaseModule"
```
But I got the following param mismatch

```
  File "eval.py", line 16, in <module>
    main()
  File "/home/chl/.cache/pypoetry/virtualenvs/torch-points3d-MFvmJRe_-py3.8/lib/python3.8/site-packages/hydra/main.py", line 20, in decorated_main
    run_hydra(
  File "/home/chl/.cache/pypoetry/virtualenvs/torch-points3d-MFvmJRe_-py3.8/lib/python3.8/site-packages/hydra/_internal/utils.py", line 171, in run_hydra
    hydra.run(
  File "/home/chl/.cache/pypoetry/virtualenvs/torch-points3d-MFvmJRe_-py3.8/lib/python3.8/site-packages/hydra/_internal/hydra.py", line 82, in run
    return run_job(
  File "/home/chl/.cache/pypoetry/virtualenvs/torch-points3d-MFvmJRe_-py3.8/lib/python3.8/site-packages/hydra/plugins/common/utils.py", line 109, in run_job
    ret.return_value = task_function(task_cfg)
  File "eval.py", line 11, in main
    trainer = Trainer(cfg)
  File "/home/chl/Software/torch-points3d/torch_points3d/trainer.py", line 44, in __init__
    self._initialize_trainer()
  File "/home/chl/Software/torch-points3d/torch_points3d/trainer.py", line 85, in _initialize_trainer
    self._model: BaseModel = self._checkpoint.create_model(
  File "/home/chl/Software/torch-points3d/torch_points3d/metrics/model_checkpoint.py", line 188, in create_model
    self._initialize_model(model, weight_name)
  File "/home/chl/Software/torch-points3d/torch_points3d/metrics/model_checkpoint.py", line 240, in _initialize_model
    model.load_state_dict(state_dict, strict=False)
  File "/home/chl/.cache/pypoetry/virtualenvs/torch-points3d-MFvmJRe_-py3.8/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1044, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for PointNet2_D:
	size mismatch for FC_layer.0.0.weight: copying a param with shape torch.Size([128, 129, 1]) from checkpoint, the shape in current model is torch.Size([128, 130, 1]).


```

Do you have an idea what is happening @nicolas-chaulet? For me, it looks like the model is somehow saved incorrectly, but I just started to look into the repository.

**Comment by malte2408** [comment]
Did you find a solution for the problem? I'm experiencing the same when using the forward_scripts/forward.py as described in the README.md.

**Comment by WeixiaoGao** [comment]
I have the same problem. And I've tried to debug and I've found out if comment out the "kwargs.pop(name)" in "torch_points3d\models\base_architectures\unet.py, line 242" as follows, it works for me. I'm not sure if it's a bug or not.
    def get_from_kwargs(self, kwargs, name):
        module = kwargs[name]
        #kwargs.pop(name)
        return module

**Comment by nicolas-chaulet** [comment]
I can see that indeed... Trying to fix it the right way, I should have something ready soon. But for now @WeixiaoGao is a good solution, thanks!

**Comment by OliverColeman** [comment]
I'm still getting the error reported by @christianlandgraf (`size mismatch for FC_layer.0.0.weight: copying a param with shape torch.Size([128, 129, 1]) from checkpoint, the shape in current model is torch.Size([128, 130, 1]).`)

I have the latest code (I checked it includes the commit "Fix #480"). I also tried the workaround posted by @WeixiaoGao but it made no difference.

The train and forward scripts were run as listed in the README. My forward config.yaml is:
```
num_workers: 2
batch_size: 16
cuda: 0
weight_name: "miou" # Used during resume, select with model to load from [miou, macc, acc..., latest]
enable_cudnn: True
checkpoint_dir: "/workspace/code/torch-points3d/outputs/benchmark/benchmark-pointnet2_charlesssg-20210202_102042" # "{your_path}/outputs/2020-01-28/11-04-13" for example
model_name: pointnet2_charlesssg
enable_dropout: False
output_path: "/workspace/code/torch-points3d/outputs/benchmark/inf1" # Where the output goes
input_path: "/workspace/code/torch-points3d/data/shapenet/raw/02954340" # Folder where to find the data

# Dataset specific
defaults:
    - dataset: ""
      optional: True
```
